test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio;
+
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.transport.nio.channel.NioChannel;
+
+import java.io.IOException;
+import java.nio.channels.ClosedSelectorException;
+import java.nio.channels.Selector;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.CountDownLatch;
+
+public abstract class ESSelector {
+
+    protected static final int NOT_STARTED = 0;
+    protected static final int RUNNING = 1;
+    protected static final int STOPPED = 2;
+
+    protected Selector selector;
+    protected volatile int state = 0;
+    protected final ConcurrentLinkedQueue<NioChannel> channelsToClose = new ConcurrentLinkedQueue<>();
+    protected final Set<NioChannel> registeredChannels = Collections.newSetFromMap(new ConcurrentHashMap<NioChannel, Boolean>());
+
+    private final EventHandler eventHandler;
+    private final BigArrays bigArrays;
+    private volatile Thread thread;
+    private CountDownLatch shutdownLatch = new CountDownLatch(1);
```

Comment:
it's not trivial :) I took a look at it and here is a suggestion, lemme know what you think:

```diff
diff --git a/test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java b/test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
index 22d9193d8e..a88dd280ae 100644
--- a/test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
+++ b/test/framework/src/main/java/org/elasticsearch/transport/nio/ESSelector.java
@@ -19,7 +19,6 @@
 
 package org.elasticsearch.transport.nio;
 
-import org.elasticsearch.ElasticsearchException;
 import org.elasticsearch.common.util.BigArrays;
 import org.elasticsearch.transport.nio.channel.NioChannel;
 
@@ -31,23 +30,21 @@ import java.util.Collections;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
-import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantLock;
 
 public abstract class ESSelector implements Closeable {
 
-    protected static final int NOT_STARTED = 0;
-    protected static final int RUNNING = 1;
-    protected static final int STOPPED = 2;
 
     protected Selector selector;
-    protected volatile int state = 0;
+    protected final AtomicBoolean isClosed = new AtomicBoolean(false);
     protected final ConcurrentLinkedQueue<NioChannel> channelsToClose = new ConcurrentLinkedQueue<>();
     protected final Set<NioChannel> registeredChannels = Collections.newSetFromMap(new ConcurrentHashMap<NioChannel, Boolean>());
 
     private final EventHandler eventHandler;
     private final BigArrays bigArrays;
     private volatile Thread thread;
-    private CountDownLatch shutdownLatch = new CountDownLatch(1);
+    private final ReentrantLock runLock = new ReentrantLock();
     protected ESSelector(EventHandler eventHandler, BigArrays bigArrays) throws IOException {
         this(eventHandler, bigArrays, Selector.open());
     }
@@ -59,18 +56,28 @@ public abstract class ESSelector implements Closeable {
     }
 
     public void runLoop() {
-        setThread();
-        state = RUNNING;
-        try {
-            while (state == RUNNING) {
-                singleLoop();
-            }
-        } finally {
+        if (runLock.tryLock()) {
             try {
-                cleanup();
+                ensureOpen();
+                setThread();
+                while (isClosed.get() == false) {
+                    singleLoop();
+                }
             } finally {
-                shutdownLatch.countDown();
+                try {
+                    cleanup();
+                } finally {
+                    runLock.unlock();
+                }
             }
+        } else {
+            throw new IllegalStateException("selector is already running");
+        }
+    }
+
+    private void ensureOpen() {
+        if (isClosed.get()) {
+            throw new IllegalStateException("selector is already closed");
         }
     }
 
@@ -79,7 +86,7 @@ public abstract class ESSelector implements Closeable {
             closePendingChannels();
             doSelect(300);
         } catch (ClosedSelectorException e) {
-            if (state == RUNNING) {
+            if (isRunning()) {
                 throw e;
             }
         } catch (IOException e) {
@@ -114,19 +121,12 @@ public abstract class ESSelector implements Closeable {
     }
 
     public void close(boolean shouldInterrupt) throws IOException {
-        int currentState = this.state;
-        state = STOPPED;
-        selector.close();
-        if (shouldInterrupt && thread != null) {
-            thread.interrupt();
-        }
-        if (currentState == NOT_STARTED) {
-            shutdownLatch.countDown();
-        }
-        try {
-            shutdownLatch.await();
-        } catch (InterruptedException e) {
-            throw new ElasticsearchException(e);
+        if (isClosed.compareAndSet(false, true)) {
+            selector.close();
+            if (shouldInterrupt && thread != null) {
+                thread.interrupt();
+            }
+            runLock.lock(); // wait for the shutdown to complete
         }
     }
 
@@ -163,6 +163,6 @@ public abstract class ESSelector implements Closeable {
     }
 
     public boolean isRunning() {
-        return state == RUNNING;
+        return runLock.isLocked();
     }
 }
```